clang/lib/Parse/ParseTemplate.cpp

```diff
@@ -960,6 +960,8 @@ Parser::ParseNonTypeTemplateParameter(unsigned Depth, unsigned Position) {
       EnterExpressionEvaluationContext ConstantEvaluated(
           Actions, Sema::ExpressionEvaluationContext::ConstantEvaluated);
       DefaultArg = Actions.CorrectDelayedTyposInExpr(ParseInitializer());
+      if (DefaultArg.isUsable())
+        DefaultArg = Actions.ActOnConstantExpression(DefaultArg);
```

Comment:
How about
```cpp
DefaultArg = Actions.ActOnConstantExpression(ParseInitializer())
```

Besides, I still don't understand why we shouldn't accept such code, according to the approach taken here
```cpp
struct S {};

void foo()
{
    constexpr S x{};
    auto a = []<auto = x>{};
}
```

but rather accept only non-record-type cases
```cpp
void foo()
{
    constexpr int x{};
    auto a = []<auto = x>{};
}
```

Can you add some standard quotes here?

Reply:
I am not sure what you mean. both should be valid

Reply:
If you look at `Sema::CheckLValueToRValueConversionOperand()`, it would bail out if `E` is of a RecordType.
So the following would still be rejected
```cpp
struct S {};

void foo()
{
    constexpr S x{};
    auto a = []<auto = x>{};
}
```

Reply:
Oh, sorry. Yes
https://eel.is/c++draft/basic#def.odr-5

Reply:
https://compiler-explorer.com/z/Ybn4nbaz4 is another example of that where compilers agree

Reply:
OK, I'm sold. (The *intervening scope* concept is still a bit harder for me to understand at the moment, so I'll probably look into it at some point.)

Feel free to merge this after one nit, which makes it a bit simpler

```cpp
DefaultArg = Actions.ActOnConstantExpression(ParseInitializer());
```

It would also be great if you could add the record-type case to the test. :)